internal/controller/ngrok/cloudendpoint_controller.go

```diff
@@ -0,0 +1,342 @@
+}
+
+func (r *CloudEndpointReconciler) updateStatus(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint, ngrokClep *ngrok.Endpoint) error {
+	clep.Status.ID = ngrokClep.ID
+	return r.Client.Status().Update(ctx, clep)
+}
+
+// #region Helper Functions
+
+// findCloudEndpointForTrafficPolicy searches for any Cloud Endpoints CRs that have a reference to a particular Traffic Policy
+func (r *CloudEndpointReconciler) findCloudEndpointForTrafficPolicy(ctx context.Context, o client.Object) []ctrl.Request {
+	tp, ok := o.(*ngrokv1alpha1.NgrokTrafficPolicy)
+	if !ok {
+		return nil
+	}
+
+	// Use the index to find CloudEndpoints that reference this TrafficPolicy
+	var cloudEndpointList ngrokv1alpha1.CloudEndpointList
+	if err := r.Client.List(ctx, &cloudEndpointList,
+		client.InNamespace(tp.Namespace),
+		client.MatchingFields{trafficPolicyNameIndex: tp.Name}); err != nil {
+		r.Log.Error(err, "failed to list CloudEndpoints using index")
+		return nil
+	}
+
+	// Collect the requests for matching CloudEndpoints
+	var requests []ctrl.Request
+	for _, clep := range cloudEndpointList.Items {
+		requests = append(requests, ctrl.Request{
+			NamespacedName: client.ObjectKey{
+				Name:      clep.Name,
+				Namespace: clep.Namespace,
+			},
+		})
+	}
+
+	return requests
+}
+
+// findTrafficPolicy fetches the TrafficPolicy CRD from the API server and returns the JSON policy as a string
+func (r *CloudEndpointReconciler) findTrafficPolicy(ctx context.Context, tpName, tpNamespace string) (string, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("name", tpName, "namespace", tpNamespace)
+
+	// Create a TrafficPolicy object to store the fetched result
+	tp := &ngrokv1alpha1.NgrokTrafficPolicy{}
+	key := client.ObjectKey{Name: tpName, Namespace: tpNamespace}
+
+	// Attempt to get the TrafficPolicy from the API server
+	if err := r.Client.Get(ctx, key, tp); err != nil {
+		log.Error(err, "failed to find TrafficPolicy CRD")
+		return "", err
+	}
+
+	// Convert the JSON policy to a string
+	policyBytes, err := tp.Spec.Policy.MarshalJSON()
+	if err != nil {
+		log.Error(err, "failed to marshal TrafficPolicy JSON")
+		return "", err
+	}
+
+	return string(policyBytes), nil
+}
+
+// ensureDomainExists checks if the Domain CRD exists, and if not, creates it.
+func (r *CloudEndpointReconciler) ensureDomainExists(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint) error {
+	domain := r.extractDomain(clep)
+	if domainEndsInReservedTLD(domain) {
+		// Skip creating the Domain CRD for reserved TLDs
+		return nil
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("domain", domain)
+
+	// Check if the Domain CRD already exists
+	domainObj := &ingressv1alpha1.Domain{}
+	err := r.Get(ctx, client.ObjectKey{Name: domain, Namespace: clep.Namespace}, domainObj)
+	if err == nil {
+		// Domain already exists
+		return nil
+	}
+	if client.IgnoreNotFound(err) != nil {
+		// Some other error occurred
+		log.Error(err, "failed to check Domain CRD existence")
+		return err
+	}
```

Comment:
Nit: One thing I notice in this controller is that we don't reach for `r.Recorder.Event()` very often. What kinds of events do you think this controller's resources would like to have?

Reply:
i think the base controller handles most of the crud events. For this controller I wanted to make sure to handle the domain creation properly so its events look like this

```
3m40s       Warning   CreateError         cloudendpoint/my-awesome-endpoint                                  Failed to Create CloudEndpoint/my-awesome-endpoint: domain is being created, requeue after delay
3m40s       Normal    Creating            domain/my-operator-based-cloud-endpoint-alexbezek-test.ngrok.io    Creating Domain/my-operator-based-cloud-endpoint-alexbezek-test.ngrok.io
3m40s       Normal    Updated             domain/my-operator-based-cloud-endpoint-alexbezek-test.ngrok.io    Updating Domain my-operator-based-cloud-endpoint-alexbezek-test.ngrok.io
3m40s       Normal    Created             domain/my-operator-based-cloud-endpoint-alexbezek-test.ngrok.io    Created Domain/my-operator-based-cloud-endpoint-alexbezek-test.ngrok.io
3m30s       Normal    Created             cloudendpoint/my-awesome-endpoint                                  Created CloudEndpoint/my-awesome-endpoint
```

Otherwise the only unique thing is the reference to traffic policy. Are there other things you think we should consider eventing on?

Reply:
Sometimes I think parsing errors or other invalid states are good to event (client/configuration errors mostly) because I find I reach for `kubectl describe` as a debugging tool before `kubectl logs` on the controller's Pod.

Reply:
hmm yeah, probably parsing out the url could have some events if its maformed